cli/cmdmanager/manager.go

```diff
@@ -0,0 +1,410 @@
+	m := CommandManager{
+		cli: cli,
+	}
+
+	// create an entity for using with command asset filtering
+	systemInfo, err := system.Info()
+	if err != nil {
+		return nil, err
+	}
+	meta := corev2.NewObjectMeta("sensuctl", "")
+	entity := &corev2.Entity{
+		EntityClass: "sensuctl",
+		ObjectMeta:  meta,
+		System:      systemInfo,
+	}
+
+	cacheDir := path.UserCacheDir("sensuctl")
+	m.db, err = bolt.Open(filepath.Join(cacheDir, dbName), 0600, &bolt.Options{
+		Timeout: 60 * time.Second,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// start the asset manager
+	ctx := context.TODO()
+	wg := sync.WaitGroup{}
+	m.assetManager = asset.NewManager(cacheDir, entity, &wg)
+	m.assetGetter, err = m.assetManager.StartAssetManager(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+func (m *CommandManager) InstallCommandFromBonsai(alias, bonsaiAssetName string) error {
+	bAsset, err := corev2.NewBonsaiBaseAsset(bonsaiAssetName)
+	if err != nil {
+		return err
+	}
+
+	var version *goversion.Version
+	if bAsset.Version != "" {
+		version, err = goversion.NewVersion(bAsset.Version)
+		if err != nil {
+			return err
+		}
+	}
+
+	bonsaiClient := bonsai.New(bonsai.BonsaiConfig{})
+	bonsaiAsset, err := bonsaiClient.FetchAsset(bAsset.Namespace, bAsset.Name)
+	if err != nil {
+		return err
+	}
+
+	if version == nil {
+		fmt.Println("no version specified, using latest:", bonsaiAsset.LatestVersion())
+		version = bonsaiAsset.LatestVersion()
+	} else if !bonsaiAsset.HasVersion(version) {
+		availableVersions := bonsaiAsset.ValidVersions()
+		sort.Sort(goversion.Collection(availableVersions))
+		availableVersionStrs := []string{}
+		for _, v := range availableVersions {
+			availableVersionStrs = append(availableVersionStrs, v.String())
+		}
+		return fmt.Errorf("version \"%s\" of asset \"%s/%s\" does not exist\navailable versions: %s",
+			version, bAsset.Namespace, bAsset.Name, strings.Join(availableVersionStrs, ", "))
+	}
+
+	fmt.Printf("fetching bonsai asset: %s/%s:%s\n", bAsset.Namespace, bAsset.Name, version)
+
+	assetJSON, err := bonsaiClient.FetchAssetVersion(bAsset.Namespace, bAsset.Name, version.String())
+	if err != nil {
+		return err
+	}
+
+	var asset corev2.Asset
+	if err := json.Unmarshal([]byte(assetJSON), &asset); err != nil {
+		return err
+	}
+
+	if val, ok := asset.Annotations["io.sensu.bonsai.type"]; ok {
+		if val != "sensuctl" {
+			return errors.New("requested asset is not a sensuctl asset")
+		}
+	} else {
+		return errors.New("requested asset does not have a type annotation set")
+	}
+
+	if val, ok := asset.Annotations["io.sensu.bonsai.provider"]; ok {
+		if val != "sensuctl/command" {
+			return errors.New("requested asset is not a sensuctl/command asset")
+		}
+	} else {
+		return errors.New("requested asset does not have a provider annotation set")
+	}
+
+	return m.installCommand(alias, &asset)
+}
+
+func (m *CommandManager) InstallCommandFromURL(alias, assetURL, checksum string) error {
```

Comment:
I must admit I found the UX of this subcommand a bit weird. Basically you need to link to an URL that points to an asset resource definition, and you also need to specify the checksum of that definition.

I took a look at the specs for this feature and it seems that when using an URL, we should actually point to the _tar.gz_ archive: https://docs.google.com/document/d/1veQ30ONMDFfZOayZLpAjmzgr1V96ahUK62uWWnaQiU4/edit#bookmark=id.yn2wm05esopd. That being said, maybe the specs changed but if so, I found that really strange.

Additionally, could we also support a local archive? I feel like it could be a quick win.

Reply:
This is a misunderstanding on my part. I think we can change this to point to a single archive but we'll lose the ability to install the correct archive based on what platform sensuctl is running on.

Reply:
Yeah, I think it's fine IMO; that's an added value of Bonsai, but it still let users easily install command plugins locally or from a private repo